### ISSUE TITLE
Logging changes folders when doing roll over. 

### ISSUE THREAD

**Issue by SWvheerden** [tari-project/tari#1845]
We need to discuss potential fixes to this. Currently the default logging changes folders when it does a roll over log. The logging is completely controlled by what happens in` log4rs.yml`, all of the behaviour is in this, what log files are created, detail of logs and where the log.  

But the current behaviour in code is to change the working directory to the one specified by the base_node data directory. The default setting in ` log4rs.yml` is a relative path of `log/*` which means the the first log (since startup) is then in `{Base_node data dir}/log/*`. But as soon as the logs roll over, the logs go to `{working dir}/log/*`. This all happens because the path's in ` log4rs.yml` is by default defined as `log\*` and this is customisable. 

So the lets discuss solutions:
# Modify default yml file
We can only modify the yml on install of new one(copying over the default sample over to use). Modify will be to insert an absolute path to the config as the default to ensure that the location remains the same.  This will only do it if the yml file is not found. If an existing file is found, the file is not edited.

## Builder pattern
The log4rs files can be created by a builder pattern, this will give us the simplest code based solution to implement. Currently we use a `log4rs-sample.yml` file as the default template. Butr we can recreate this in a builder pattern to create a new one with correct absolute paths relative to what the user set as the `Base_node_working dir` on first start up. 

This method will be harder to edit the default as the user needs to edit rust code to change the defaults. 

## Yml editor
The other solution is to continue to use `log4rs-sample.yml` as the default, but as the user starts up the node for the first time, the code will read and edit the yml file to change the paths to the correct absolute paths relative to what the user set as the `Base_node_working dir` on first start up.

This method will be easier to edit the default logging settings, but it will add code complexity and perhaps extra dependancies. 

# Script based
 We can create startup scripts which ensure that when a user starts the base node the `Working dir` is correctly set and the base nodes should then log correct. The default way of staring the base then will be via a script and not a binary. 

This is the easiest and least complex method to implement. 

Am I missing solutions?

**Comment by hansieodendaal** [comment]
It seems the **log4rs** guys have been busy with similar issues, see [issue](https://github.com/estk/log4rs/issues/21) and related [PR](https://github.com/estk/log4rs/pull/155). It is currently merged into their development branch and should be merged into their master branch and tagged when they are happy with it.

**Comment by SWvheerden** [comment]
But it still does not solve our issue, we dont currently by default use environment variables. But it does give us more options to play with. 

If its merged into master, we cause force env_var to be used by default

**Comment by dunnock** [comment]
Usually logs are going to predefined directory so fixing the path on the first start sounds like a good solutions. Described `Yml editor` approach is quite common and should be clear for users.

BTW, did we consider changing log format to toml? There is `toml` feature which allows log4rs reading config from toml files.

**Comment by hansieodendaal** [comment]
I prefer starting `tari_base_node` with a script so I can control the environment and command-line arguments. I make sure the script is started in the `base-path` folder, then I can use relative paths in `log4rs.yml`. Rollover always works, and the exe can be in any path.

My script for development, where I would like to keep copies of the exe, looks like this:
```
export my_exe="tari_base_node"
export target_type="release"
export build_folder="${HOME}/tari-project/target/${target_type}"
export config_path="./config"
echo

echo \>\> Get base node executable
cp $my_exe ${my_exe}_${target_type}_`date +%d%m%y-%H%M%S`
cp ${build_folder}/${my_exe} $my_exe
echo

echo \>\> Create ${my_exe} configuration data
if [ ! -f "${config_path}/node_id.json" ] 
then
    echo
    ./$my_exe --create_id --config "${config_path}/config.toml" --log_config "${config_path}/log4rs.yml" --base-path .
else
    echo Using old "${config_path}/config.toml";
fi
echo

./$my_exe --config "${config_path}/config.toml" --log_config "${config_path}/log4rs.yml" --base-path .
```

**Comment by hansieodendaal** [comment]
> Usually logs are going to predefined directory so fixing the path on the first start sounds like a good solutions. Described `Yml editor` approach is quite common and should be clear for users.

If we are going to do this I suggest having a command-line parameter that makes it clear logging will be to an absolute path and not a relative path. With a relative path you can change the base test folder's name and everything will still work as expected.

**Comment by CjS77** [comment]
I'm somewhat confused:
> But the current behaviour in code is to change the working directory to the one specified by the base_node data directory. 

Where is logging output controlled in the code? (There should only ever be `log::*` macro invocations.) Or do you mean the bug in the rolling file appender implementation? 

As to the question around default logging configuration, I still favour the current approach. Keeping all config separate from code beats this issue IMO. So if there's a way to work around the bug by only changing the default `log4rs.yml` file, then let's do that.

IF you want/need to edit `log4rs.yml` dynamically, I suggest don't do it in the `tari_base_node` executable, but in the installation script(s) instead. The reason is that I'm always fiddling with the log4rs.yml file to tweak things, point to different folders etc., and I'd be pissed if the executable kept overwriting my changes.

Recommended reading:  https://12factor.net/config as guidance

**Comment by SWvheerden** [comment]
> e are going to do this I suggest having a command-line parameter that makes it clear logging will be to an absolute path and not a relative path. With a relative path you can change th

But the problem with this is, the current default, with reletive, only works correctly if you start up the base_node via a script with the `working dir` set correctly.  I might be wrong but I think a minority of users start the node via a script

**Comment by SWvheerden** [comment]
> I'm somewhat confused:
> 
> > But the current behaviour in code is to change the working directory to the one specified by the base_node data directory.
> 
> Where is logging output controlled in the code? (There should only ever be `log::*` macro invocations.) Or do you mean the bug in the rolling file appender implementation?
> 
> As to the question around default logging configuration, I still favour the current approach. Keeping all config separate from code beats this issue IMO. So if there's a way to work around the bug by only changing the default `log4rs.yml` file, then let's do that.
> 
> IF you want/need to edit `log4rs.yml` dynamically, I suggest don't do it in the `tari_base_node` executable, but in the installation script(s) instead. The reason is that I'm always fiddling with the log4rs.yml file to tweak things, point to different folders etc., and I'd be pissed if the executable kept overwriting my changes.
> 
> Recommended reading: https://12factor.net/config as guidance

https://github.com/tari-project/tari/blob/10faadca9b40c9bde98882cd70274c796b4e00d7/common/src/configuration/bootstrap.rs#L177

Thats where the directory is changed for logging. 


Also the proposal to edit is only for the startup if no yml file is found. This was not explicitly stated, but its updated now.

**Comment by philipr-za** [comment]
Some more context: currently on Mac if you use the installer it copies the binary into `/usr/local/bin` and makes it available in the global path so the user can run that binary from any working directory and when the log4rs files rollover the new files end up in a log folder relative to the working directory. 

@hansieodendaal In the Mac and Unix environments we have standardised on using the ~/.tari folder as the location for the databases and config files, it feels like that is where the logs should be. Unfortunately log4rs does not understand the ~ as home in the path. 

I also think that requiring the binary to be started using a script is not ideal. You can if you have specific needs but really the default behaviour of just running the binary should be the most common setup we want to support and for unix environments it feels like adding it to the global bin directory is something we should support.

We should definitiely not edit any config files after they have been deployed during an install. My vote would be that during the step where if the binary is started and a logging config file does not exist in ~/.tari, when we copy the default config file (log4rs-sample.yml) into ~/.tari at that stage we update the logging paths in the resulting config file to be absolute paths to ~/.tari e.g. /Users/philip/.tari 

We should then remove that line mentioned in the comment above so that the config file fully defines where the logs end up.

**Comment by CjS77** [comment]
It's pretty common to use scripts as "the" executable when installing things from packages. Users wouldn't even know they're running a bootstrap script.

In OSX and Linux, one could make `tari_base_node`(a bash script) the executable, install it in `/usr/local/bin` or equivalent, perform any pre-execution env setup and then call the binary itself (installed in `/usr/local/tari` for instance) carrying over any CLI arguments

**Comment by philipr-za** [comment]
Ok so we could use a bash script in `/usr/local/bin`to force the working directory to `~/.tari` before running the binary. Though in that case it is not going to be clear where the path for the logs defined in the `~/.tari/log4rs.yml` is pointing? 

If we rely on scripts to manage these working directories we will also have to update the build/install process in Rust for those that build from source to get consistent behaviour.

**Comment by hansieodendaal** [comment]
@philipr-za, `~/.tari` is the default `--base-path` for the database and config files, if the user does not override it with another `--base-path` on the command line. I agree with the logical conclusion that the log files should end up in `--base-path`. Currently with merging of [PR#1749](https://github.com/tari-project/tari/pull/1749) the code was changed to also initialize the log files in `--base-path` and not in the current working dir.

So the only requirement currently for rollover is to make sure the exe or script calling the exe is started in `--base-path`. When `log4rs` implements [PR#155](https://github.com/estk/log4rs/pull/155) we should be able to have more control of rollover files via environment variables.

**Comment by SWvheerden** [comment]
Does 155 give us access to one var to the home/user dir for windows/mac/linux?

**Comment by philipr-za** [comment]
Don't think we should wait on log4rs to resolve this.

So I can see that baking anything into the binary has an issue for being cross platform. Currently we do bake this config process into the binary though. So feels like we should manage this process in the install scripts and make sure that the install script puts the correct log4rs.yml into the correct place and potentially dynamically edits the paths in it to point to the correct location.

We should then be careful to document or automate this process for anyone who builds the binary from source because if they build the binary and just run it then it will be borked as we need the install and/or run scripts to have it work correctly?

**Comment by hansieodendaal** [comment]
Seeing that roll-over does not work as expected, I think this could be resolved by just reverting a portion of #1749 to not initialize logging in `--base-path`, but in the current working dir. So `cargo run --bin ...` will create `~/.tari` as the default `--base-path` for the db and config files and `./log` for the log files. If the exe is started directly or a script is used to start the exe from `--base-path` then `./log` will also be in `--base-path`.

When `log4rs` implements [PR#155](https://github.com/estk/log4rs/pull/155) we can add additional environment variables.

Votes anyone? Then I can quickly do this.

**Comment by philipr-za** [comment]
I am fine with that for now, just so that the behaviour is consistent.

**Comment by hansieodendaal** [comment]
See #1853

**Comment by Tas4tari** [comment]
As per @hansieodendaal this has been fixed and the PR commented above has been merged therefore closing this issue
